File: Original/Bridge/Bridge/Resources/.generated/System/Text/UTF8Encoding.js
    Bridge.define("System.Text.UTF8Encoding", {
        inherits: [System.Text.Encoding],
        fields: {
            encoderShouldEmitUTF8Identifier: false,
            throwOnInvalid: false
        },
        props: {
            CodePage: {
                get: function () {
                    return 65001;
                }
            },
            EncodingName: {
                get: function () {
                    return "Unicode (UTF-8)";
                }
            }
        },
        ctors: {
            ctor: function () {
                System.Text.UTF8Encoding.$ctor1.call(this, false);
            },
            $ctor1: function (encoderShouldEmitUTF8Identifier) {
                System.Text.UTF8Encoding.$ctor2.call(this, encoderShouldEmitUTF8Identifier, false);
            },
            $ctor2: function (encoderShouldEmitUTF8Identifier, throwOnInvalidBytes) {
                this.$initialize();
                System.Text.Encoding.ctor.call(this);
                this.encoderShouldEmitUTF8Identifier = encoderShouldEmitUTF8Identifier;
                this.throwOnInvalid = throwOnInvalidBytes;
                this.fallbackCharacter = 65533;
            }
        },
        methods: {
            Encode$3: function (s, outputBytes, outputIndex, writtenBytes) {
                var hasBuffer = outputBytes != null;
                var record = 0;

                var write = function (args) {
                    var len = args.length;
                    for (var j = 0; j < len; j = (j + 1) | 0) {
                        var code = args[System.Array.index(j, args)];
                        if (hasBuffer) {
                            if (outputIndex >= outputBytes.length) {
                                throw new System.ArgumentException.$ctor1("bytes");
                            }

                            outputBytes[System.Array.index(Bridge.identity(outputIndex, ((outputIndex = (outputIndex + 1) | 0))), outputBytes)] = code;
                        } else {
                            outputBytes.push(code);
                        }
                        record = (record + 1) | 0;
                    }
                };

                var fallback = Bridge.fn.bind(this, $asm.$.System.Text.UTF8Encoding.f1);

                if (!hasBuffer) {
                    outputBytes = System.Array.init(0, 0, System.Byte);
                }

                for (var i = 0; i < s.length; i = (i + 1) | 0) {
                    var charcode = s.charCodeAt(i);

                    if ((charcode >= 55296) && (charcode <= 56319)) {
                        var next = s.charCodeAt(((i + 1) | 0));
                        if (!((next >= 56320) && (next <= 57343))) {
                            charcode = fallback();
                        }
                    } else if ((charcode >= 56320) && (charcode <= 57343)) {
                        charcode = fallback();
                    }

                    if (charcode < 128) {
                        write(System.Array.init([charcode], System.Byte));
                    } else if (charcode < 2048) {
                        write(System.Array.init([(192 | (charcode >> 6)), (128 | (charcode & 63))], System.Byte));
                    } else if (charcode < 55296 || charcode >= 57344) {
                        write(System.Array.init([(224 | (charcode >> 12)), (128 | ((charcode >> 6) & 63)), (128 | (charcode & 63))], System.Byte));
                    } else {
                        i = (i + 1) | 0;
                        var code = (65536 + (((charcode & 1023) << 10) | (s.charCodeAt(i) & 1023))) | 0;
                        write(System.Array.init([(240 | (code >> 18)), (128 | ((code >> 12) & 63)), (128 | ((code >> 6) & 63)), (128 | (code & 63))], System.Byte));
                    }
                }

                writtenBytes.v = record;

                if (hasBuffer) {
                    return null;
                }

                return outputBytes;
            },
            Decode$2: function (bytes, index, count, chars, charIndex) {
                this._hasError = false;
                var position = index;
                var result = "";
                var surrogate1 = 0;
                var addFallback = false;
                var endpoint = (position + count) | 0;

                for (; position < endpoint; position = (position + 1) | 0) {
                    var accumulator = 0;
                    var extraBytes = 0;
                    var hasError = false;
                    var firstByte = bytes[System.Array.index(position, bytes)];

                    if (firstByte <= 127) {
                        accumulator = firstByte;
                    } else if ((firstByte & 64) === 0) {
                        hasError = true;
                    } else if ((firstByte & 224) === 192) {
                        accumulator = firstByte & 31;
                        extraBytes = 1;
                    } else if ((firstByte & 240) === 224) {
                        accumulator = firstByte & 15;
                        extraBytes = 2;
                    } else if ((firstByte & 248) === 240) {
                        accumulator = firstByte & 7;
                        extraBytes = 3;
                    } else if ((firstByte & 252) === 248) {
                        accumulator = firstByte & 3;
                        extraBytes = 4;
                        hasError = true;
                    } else if ((firstByte & 254) === 252) {
                        accumulator = firstByte & 3;
                        extraBytes = 5;
                        hasError = true;
                    } else {
                        accumulator = firstByte;
                        hasError = false;
                    }

                    while (extraBytes > 0) {
                        position = (position + 1) | 0;

                        if (position >= endpoint) {
                            hasError = true;
                            break;
                        }

                        var extraByte = bytes[System.Array.index(position, bytes)];
                        extraBytes = (extraBytes - 1) | 0;

                        if ((extraByte & 192) !== 128) {
                            position = (position - 1) | 0;
                            hasError = true;
                            break;
                        }

                        accumulator = (accumulator << 6) | (extraByte & 63);
                    }

                    /* if ((accumulator == 0xFFFE) || (accumulator == 0xFFFF)) {
                       hasError = true;
                    }*/

                    var characters = null;
                    addFallback = false;
                    if (!hasError) {
                        if (surrogate1 > 0 && !((accumulator >= 56320) && (accumulator <= 57343))) {
                            hasError = true;
                            surrogate1 = 0;
                        } else if ((accumulator >= 55296) && (accumulator <= 56319)) {
                            surrogate1 = accumulator & 65535;
                        } else if ((accumulator >= 56320) && (accumulator <= 57343)) {
                            hasError = true;
                            addFallback = true;
                            surrogate1 = 0;
                        } else {
                            characters = System.String.fromCharCode(accumulator);
                            surrogate1 = 0;
                        }
                    }

                    if (hasError) {
                        if (this.throwOnInvalid) {
                            throw new System.Exception("Invalid character in UTF8 text");
                        }

                        result = (result || "") + String.fromCharCode(this.fallbackCharacter);
                        this._hasError = true;
                    } else if (surrogate1 === 0) {
                        result = (result || "") + (characters || "");
                    }
                }

                if (surrogate1 > 0 || addFallback) {
                    if (this.throwOnInvalid) {
                        throw new System.Exception("Invalid character in UTF8 text");
                    }

                    if (result.length > 0 && result.charCodeAt(((result.length - 1) | 0)) === this.fallbackCharacter) {
                        result = (result || "") + String.fromCharCode(this.fallbackCharacter);
                    } else {
                        result = (result || "") + (((this.fallbackCharacter + this.fallbackCharacter) | 0));
                    }

                    this._hasError = true;
                }

                return result;
            },
            GetMaxByteCount: function (charCount) {
                if (charCount < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("charCount");
                }


                var byteCount = System.Int64(charCount).add(System.Int64(1));
                byteCount = byteCount.mul(System.Int64(3));

                if (byteCount.gt(System.Int64(2147483647))) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("charCount");
                }

                return System.Int64.clip32(byteCount);
            },
            GetMaxCharCount: function (byteCount) {
                if (byteCount < 0) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("byteCount");
                }

                var charCount = System.Int64(byteCount).add(System.Int64(1));

                if (charCount.gt(System.Int64(2147483647))) {
                    throw new System.ArgumentOutOfRangeException.$ctor1("byteCount");
                }

                return System.Int64.clip32(charCount);
            }
        }
    });

    Bridge.ns("System.Text.UTF8Encoding", $asm.$);

    Bridge.apply($asm.$.System.Text.UTF8Encoding, {
        f1: function () {
            if (this.throwOnInvalid) {
                throw new System.Exception("Invalid character in UTF8 text");
            }

            return this.fallbackCharacter;
        }
    });

File: Original/Bridge/Bridge/System/Text/UTF8Encoding.cs
namespace System.Text
{
    public class UTF8Encoding : Encoding
    {
        private bool encoderShouldEmitUTF8Identifier;
        private bool throwOnInvalid;

        public UTF8Encoding() : this(false)
        {
        }

        public UTF8Encoding(bool encoderShouldEmitUTF8Identifier) : this(encoderShouldEmitUTF8Identifier, false)
        {
        }

        public UTF8Encoding(bool encoderShouldEmitUTF8Identifier, bool throwOnInvalidBytes)
        {
            this.encoderShouldEmitUTF8Identifier = encoderShouldEmitUTF8Identifier;
            this.throwOnInvalid = throwOnInvalidBytes;
            this.fallbackCharacter = '\uFFFD';
        }

        public override int CodePage => 65001;
        public override string EncodingName => "Unicode (UTF-8)";

        protected override byte[] Encode(string s, byte[] outputBytes, int outputIndex, out int writtenBytes)
        {
            var hasBuffer = outputBytes != null;
            var record = 0;

            Action<byte[]> write = args =>
            {
                var len = args.Length;
                for (var j = 0; j < len; j++)
                {
                    var code = args[j];
                    if (hasBuffer)
                    {
                        if (outputIndex >= outputBytes.Length)
                        {
                            throw new System.ArgumentException("bytes");
                        }

                        outputBytes[outputIndex++] = code;
                    }
                    else
                    {
                        outputBytes.Push(code);
                    }
                    record++;
                }
            };

            Func<char> fallback = () =>
            {
                if (this.throwOnInvalid)
                {
                    throw new System.Exception("Invalid character in UTF8 text");
                }

                return this.fallbackCharacter;
            };

            if (!hasBuffer)
            {
                outputBytes = new byte[0];
            }

            for (var i = 0; i < s.Length; i++)
            {
                var charcode = s[i];

                if ((charcode >= 0xD800) && (charcode <= 0xDBFF))
                {
                    var next = s[i + 1];
                    if (!((next >= 0xDC00) && (next <= 0xDFFF)))
                    {
                        charcode = fallback();
                    }
                }
                else if ((charcode >= 0xDC00) && (charcode <= 0xDFFF))
                {
                    charcode = fallback();
                }

                if (charcode < 0x80)
                {
                    write(new[] { charcode.As<byte>() });
                }
                else if (charcode < 0x800)
                {
                    write(new[] { (0xc0 | (charcode >> 6)).As<byte>(), (0x80 | (charcode & 0x3f)).As<byte>() });
                }
                else if (charcode < 0xd800 || charcode >= 0xe000)
                {
                    write(new[]{(0xe0 | (charcode >> 12)).As<byte>(),
                                (0x80 | ((charcode >> 6) & 0x3f)).As<byte>(),
                                (0x80 | (charcode & 0x3f)).As<byte>()});
                }
                else
                {
                    i++;
                    var code = 0x10000 + (((charcode & 0x3ff) << 10) | (s[i] & 0x3ff));
                    write(new[]{(0xf0 | (code >> 18)).As<byte>(),
                              (0x80 | ((code >> 12) & 0x3f)).As<byte>(),
                              (0x80 | ((code >> 6) & 0x3f)).As<byte>(),
                              (0x80 | (code & 0x3f)).As<byte>()});
                }
            }

            writtenBytes = record;

            if (hasBuffer)
            {
                return null;
            }

            return outputBytes;
        }

        protected override string Decode(byte[] bytes, int index, int count, char[] chars, int charIndex)
        {
            this._hasError = false;
            var position = index;
            var result = "";
            var surrogate1 = '\u0000';
            var addFallback = false;
            var endpoint = position + count;

            for (; position < endpoint; position++)
            {
                var accumulator = 0;
                var extraBytes = 0;
                var hasError = false;
                var firstByte = bytes[position];

                if (firstByte <= 0x7F)
                {
                    accumulator = firstByte;
                }
                else if ((firstByte & 0x40) == 0)
                {
                    hasError = true;
                }
                else if ((firstByte & 0xE0) == 0xC0)
                {
                    accumulator = firstByte & 31;
                    extraBytes = 1;
                }
                else if ((firstByte & 0xF0) == 0xE0)
                {
                    accumulator = firstByte & 15;
                    extraBytes = 2;
                }
                else if ((firstByte & 0xF8) == 0xF0)
                {
                    accumulator = firstByte & 7;
                    extraBytes = 3;
                }
                else if ((firstByte & 0xFC) == 0xF8)
                {
                    accumulator = firstByte & 3;
                    extraBytes = 4;
                    hasError = true;
                }
                else if ((firstByte & 0xFE) == 0xFC)
                {
                    accumulator = firstByte & 3;
                    extraBytes = 5;
                    hasError = true;
                }
                else
                {
                    accumulator = firstByte;
                    hasError = false;
                }

                while (extraBytes > 0)
                {
                    position++;

                    if (position >= endpoint)
                    {
                        hasError = true;
                        break;
                    }

                    var extraByte = bytes[position];
                    extraBytes--;

                    if ((extraByte & 0xC0) != 0x80)
                    {
                        position--;
                        hasError = true;
                        break;
                    }

                    accumulator = (accumulator << 6) | (extraByte & 0x3F);
                }

                /*if ((accumulator == 0xFFFE) || (accumulator == 0xFFFF)) {
                    hasError = true;
                }*/

                string characters = null;
                addFallback = false;
                if (!hasError)
                {
                    if (surrogate1 > 0 && !((accumulator >= 0xDC00) && (accumulator <= 0xDFFF)))
                    {
                        hasError = true;
                        surrogate1 = '\u0000';
                    }
                    else if ((accumulator >= 0xD800) && (accumulator <= 0xDBFF))
                    {
                        surrogate1 = (char)accumulator;
                    }
                    else if ((accumulator >= 0xDC00) && (accumulator <= 0xDFFF))
                    {
                        hasError = true;
                        addFallback = true;
                        surrogate1 = '\u0000';
                    }
                    else
                    {
                        characters = Encoding.FromCharCode(accumulator);
                        surrogate1 = '\u0000';
                    }
                }

                if (hasError)
                {
                    if (this.throwOnInvalid)
                    {
                        throw new System.Exception("Invalid character in UTF8 text");
                    }

                    result += this.fallbackCharacter;
                    this._hasError = true;
                }
                else if (surrogate1 == 0)
                {
                    result += characters;
                }
            }

            if (surrogate1 > 0 || addFallback)
            {
                if (this.throwOnInvalid)
                {
                    throw new System.Exception("Invalid character in UTF8 text");
                }

                if (result.Length > 0 && result[result.Length - 1] == this.fallbackCharacter)
                {
                    result += this.fallbackCharacter;
                }
                else
                {
                    result += this.fallbackCharacter + this.fallbackCharacter;
                }

                this._hasError = true;
            }

            return result;
        }

        public override int GetMaxByteCount(int charCount)
        {
            if (charCount < 0)
            {
                throw new System.ArgumentOutOfRangeException("charCount");
            }


            var byteCount = (long)charCount + 1;
            byteCount *= 3;

            if (byteCount > 0x7fffffff)
            {
                throw new System.ArgumentOutOfRangeException("charCount");
            }

            return (int)byteCount;
        }

        public override int GetMaxCharCount(int byteCount)
        {
            if (byteCount < 0)
            {
                throw new System.ArgumentOutOfRangeException("byteCount");
            }

            var charCount = (long)byteCount + 1;

            if (charCount > 0x7fffffff)
            {
                throw new System.ArgumentOutOfRangeException("byteCount");
            }

            return (int)charCount;
        }
    }
}
